<a href="https://colab.research.google.com/github/gtobin28/Linear_Algebra_Coding_Projects/blob/master/Sports_Teams_Rankings_Grace_Tobin.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Load the preliminary code

In [1]:
import matplotlib.pyplot as plt
from sympy import Poly
import sympy as sym
import numpy as np
from numpy import c_

# We'll specify that x and t are variables:
x, t = sym.symbols('x, t')

ll = "loaded"

In [2]:
lll = " correctly!"

def rref(matrix):
    A = np.array(matrix, dtype=np.float64)
    
    r = 0 #row
    c = 0 #column

    (maxr, maxc) = A.shape

    prev_row_pivot = -1   # The previous row with a leading term:
    tmp_pivot_row = -1
    this_pivot_row = -1

    for c in range(maxc):  # go through each column
      tmp_pivot_row = -1  # reset the tmp pivot row
      this_pivot_row = -1  # reset this pivot row
      for r in range(maxr):  # Find the leading term in this column below prev_row_pivot
        if ((r > prev_row_pivot) and (A[r][c] != 0.0)):
          tmp_pivot_row = r  # find the new pivot row
      if (tmp_pivot_row > -1):  # we found a candidate for the leading term:
        #swap row this_pivot_row with (prev_row_pivot + 1)
        A[[tmp_pivot_row, prev_row_pivot+1]] = A[[prev_row_pivot + 1, tmp_pivot_row]]
        #set this_pivot_row to prev_row_pivot+1
        this_pivot_row = prev_row_pivot + 1
        #Now, normalize the row based on the pivot value:
        A[ this_pivot_row ] = A[ this_pivot_row ] / A[ this_pivot_row ][c]
        #Then, use this row to force 0's above and below the pivot entry:
        #subtract multiples of of the this_pivot_row row from the others
        for r_ in range(len(A)):
            if r_ != this_pivot_row:
                A[r_] = A[r_] - A[this_pivot_row] * A[r_][c] / A[this_pivot_row][c]
        prev_row_pivot = this_pivot_row
    return A   


def is_singular(matrix):
  singular = True
  notsingular = False

  (mr, mc) = matrix.shape
  
  return_value = notsingular
  if (mc - np.linalg.linalg.matrix_rank(matrix) > 0) :
    return_value = singular
  return return_value
 
    
    

print(">> The page ", ll, lll)

>> The page  loaded  correctly!


# Ranking Teams based on games

From our packet, we saw there were five games:

Home  | Away
------|------
Winona State 16 | Sioux Falls 15
Winona State 34 | Augustana 20
Sioux Falls 51  | Augustana 13
Duluth 28       |  Augustana 17
Winona State 17 | Duluth 37

Let's encode this system into a matrix.

The unknown rankings are $r_w, r_s, r_a, r_d$.

The system looks like $M \vec{x} = \vec{b}$, or

$$
\left( \begin{array}{cccc} 1 & -1 & 0 & 0 \\ 1 & 0 & -1 & 0 \\ 0 & 1 & -1 & 0 \\ 0 & 0 & -1 & 1 \\ -1 & 0  & 0 & 1 \\ \end{array} \right) \, \left( \begin{array}{c} r_w \\ r_s \\ r_a \\ r_d \\ \end{array} \right) = \left( \begin{array}{c} 1 \\ 14 \\ 38 \\ 11 \\ 20 \\ \end{array} \right)$$

We'll solve this system using least squares 
$$M^T I M \vec{x} = M^T I \vec{b} $$

$$ \vec{x} = \left( M^T I M  \right)^{-1} M^T I \vec{b} $$






In [3]:
# create the matrix:
M = np.matrix([[1, -1, 0, 0],
               [1, 0, -1, 0],
               [0, 1, -1, 0],
               [0, 0, -1, 1],
               [-1, 0, 0, 1]])


# create the vector for the right hand side:
pts = np.matrix([[1], [14], [38], [11], [20]])
 
#Make sure the system looks appropriate:
print(np.c_[M,pts])

[[ 1 -1  0  0  1]
 [ 1  0 -1  0 14]
 [ 0  1 -1  0 38]
 [ 0  0 -1  1 11]
 [-1  0  0  1 20]]


In [4]:
# Generate the transpose of M:
MT = M.transpose()

print("The Transpose of M is:")
print(MT)

# and find MT*M and MT*pts:
print("\n")
print("MT * M is: ")
MTM = MT * M
print(MTM)

print("\n")
print("MT * pts is:")
MTpts = MT * pts
print(MTpts)


The Transpose of M is:
[[ 1  1  0  0 -1]
 [-1  0  1  0  0]
 [ 0 -1 -1 -1  0]
 [ 0  0  0  1  1]]


MT * M is: 
[[ 3 -1 -1 -1]
 [-1  2 -1  0]
 [-1 -1  3 -1]
 [-1  0 -1  2]]


MT * pts is:
[[ -5]
 [ 37]
 [-63]
 [ 31]]


If $M^T M$ has an inverse, then we can use it.  But, if $M^T M$ is singular, then it doesn't have an inverse.  In this situation, we need to replace every entry in one row (say the bottom row) of $M^T M$ and $M^T \vec{pts}$ with 1's.  (This knocks the system out of *not* having an inverse, so it will now have an inverse.)

In [7]:
#  Check if MTM is singular
is_singular(MT*M)
#  If True, (MTM)-1 Does Not Exist& It's Singular

True

In [9]:
# If MTM is singular, then adjust one row of MTM and MT pts to be just 1's:

#MTM[3] is the entire 4th row of MTM, 
#so MTM[3] = 1 will put 1's in MTM's 4th row.

MTM[3] = 1  # Changes MTM
MTpts[3] = 1  # Changes MTpts

print("MTM is now: ")
print(MTM)

print("and MTpts is now:")
print(MTpts)

print("Is our modified MTM singular?")
is_singular(MTM)

MTM is now: 
[[ 3 -1 -1 -1]
 [-1  2 -1  0]
 [-1 -1  3 -1]
 [ 1  1  1  1]]
and MTpts is now:
[[ -5]
 [ 37]
 [-63]
 [  1]]
Is our modified MTM singular?


False

In [10]:
#Find our solution using the inverse of MTM:

#First, find the inverse of MTM:
MTMI = np.linalg.inv(MTM)

print("The inverse of MTM is:")
print(MTMI)

#Now, use the inverse of MTM to solve for the rankings:

print("\n")
print("The solution are rankings:")
MTMI * MTpts

The inverse of MTM is:
[[ 2.50000000e-01  0.00000000e+00  1.85037171e-17  2.50000000e-01]
 [ 1.25000000e-01  5.00000000e-01  1.25000000e-01  2.50000000e-01]
 [ 0.00000000e+00  0.00000000e+00  2.50000000e-01  2.50000000e-01]
 [-3.75000000e-01 -5.00000000e-01 -3.75000000e-01  2.50000000e-01]]


The solution are rankings:


matrix([[ -1.  ],
        [ 10.25],
        [-15.5 ],
        [  7.25]])

rS > rD > rW > r

If Duluth plays Sioux Falls, able to assume Sioux Falls beats Duluth by:           
[rS - rD] = [10.25 - 7.25] = 3 pts

#  Game Data

Find the game data for the four teams in the NFC North and NFC South.

The teams are:

* **NFC North:**  Green Bay Packers, Chicago Bears, Detroit Lions, Minnesota Vikings
* **NFC South:**  Atlanta Falcons, Carolina Panthers, New Orleans Saints, Tampa Bay Buccaneers


It may be helpful to do a google search to find the data.

Be sure to create your matrix on the handout, then enter the data for the system below.



In [11]:
# create the matrix:
M = np.matrix([[1, 0, 0, -1, 0, 0, 0, 0],
               [1, 0, -1, 0, 0, 0, 0, 0],
               [1, 0, 0, 0, 0, 0, -1, 0],
               [1, 0, 0, 0, -1, 0, 0, 0],
               [0, 1, 0, 0, -1, 0, 0, 0],
               [0, 1, 0, 0, 0, 0, 0, -1],
               [0, 1, -1, 0, 0, 0, 0, 0],
               [0, 0, -1, 0, 0, 0, 1, 0],
               [0, 0, 0, 0, -1, 1, 0, 0],
               [0, 0, 0, 0, 0, -1, 0, 1],
               [0, 0, 0, 0, 0, 0, 1, -1]])


# create the vector for the right hand side:
pts = np.matrix([[9], [21], [7], [14], [4], [1], [4], [6], [7], [14], [11]])
 
#Make sure the system looks appropriate:
print(np.c_[M,pts])

[[ 1  0  0 -1  0  0  0  0  9]
 [ 1  0 -1  0  0  0  0  0 21]
 [ 1  0  0  0  0  0 -1  0  7]
 [ 1  0  0  0 -1  0  0  0 14]
 [ 0  1  0  0 -1  0  0  0  4]
 [ 0  1  0  0  0  0  0 -1  1]
 [ 0  1 -1  0  0  0  0  0  4]
 [ 0  0 -1  0  0  0  1  0  6]
 [ 0  0  0  0 -1  1  0  0  7]
 [ 0  0  0  0  0 -1  0  1 14]
 [ 0  0  0  0  0  0  1 -1 11]]


In [12]:
# Generate the transpose of M:
MT = M.transpose()

print("The Transpose of M is:")
print(MT)

# and find MT*M and MT*pts:
print("\n")
print("MT * M is: ")
MTM = MT * M
print(MTM)

print("\n")
print("MT * pts is:")
MTpts = MT * pts
print(MTpts)


The Transpose of M is:
[[ 1  1  1  1  0  0  0  0  0  0  0]
 [ 0  0  0  0  1  1  1  0  0  0  0]
 [ 0 -1  0  0  0  0 -1 -1  0  0  0]
 [-1  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0 -1 -1  0  0  0 -1  0  0]
 [ 0  0  0  0  0  0  0  0  1 -1  0]
 [ 0  0 -1  0  0  0  0  1  0  0  1]
 [ 0  0  0  0  0 -1  0  0  0  1 -1]]


MT * M is: 
[[ 4  0 -1 -1 -1  0 -1  0]
 [ 0  3 -1  0 -1  0  0 -1]
 [-1 -1  3  0  0  0 -1  0]
 [-1  0  0  1  0  0  0  0]
 [-1 -1  0  0  3 -1  0  0]
 [ 0  0  0  0 -1  2  0 -1]
 [-1  0 -1  0  0  0  3 -1]
 [ 0 -1  0  0  0 -1 -1  3]]


MT * pts is:
[[ 51]
 [  9]
 [-31]
 [ -9]
 [-25]
 [ -7]
 [ 10]
 [  2]]


In [13]:
#  Check if MTM is singular
is_singular(MT*M)

True

In [16]:
# Careful - you might not need to do this!  
# Be sure to check whether MTM is singular on the previous line.

# If MTM is singular, then adjust one row of MTM and MT pts to be just 1's:

#MTM[3] is the entire 4th row of MTM, 
#so MTM[3] = 1 will put 1's in MTM's 4th row.

MTM[7] = 1
MTpts[7] = 1

print("MTM is now: ")
print(MTM)

print("and MTpts is now:")
print(MTpts)

print("Is our modified MTM singular?")
is_singular(MTM)

MTM is now: 
[[ 4  0 -1 -1 -1  0 -1  0]
 [ 0  3 -1  0 -1  0  0 -1]
 [-1 -1  3  0  0  0 -1  0]
 [-1  0  0  1  0  0  0  0]
 [-1 -1  0  0  3 -1  0  0]
 [ 0  0  0  0 -1  2  0 -1]
 [-1  0 -1  0  0  0  3 -1]
 [ 1  1  1  1  1  1  1  1]]
and MTpts is now:
[[ 51]
 [  9]
 [-31]
 [ -9]
 [-25]
 [ -7]
 [ 10]
 [  1]]
Is our modified MTM singular?


False

In [17]:
#Find our solution using the inverse of MTM:

#First, find the inverse of MTM:
MTMI = np.linalg.inv(MTM)

print("The inverse of MTM is:")
print(MTMI)

#Now, use the inverse of MTM to solve for the rankings:

print("\n")
print("The solution are rankings:")
MTMI * MTpts

The inverse of MTM is:
[[ 0.35909091  0.02727273  0.12727273  0.23409091  0.07954545 -0.02272727
   0.12045455  0.125     ]
 [-0.12272727  0.28181818 -0.01818182 -0.24772727 -0.01136364 -0.06818182
  -0.08863636  0.125     ]
 [ 0.07727273  0.08181818  0.38181818 -0.04772727 -0.01136364 -0.06818182
   0.11136364  0.125     ]
 [ 0.35909091  0.02727273  0.12727273  1.23409091  0.07954545 -0.02272727
   0.12045455  0.125     ]
 [ 0.00454545  0.06363636 -0.03636364 -0.12045455  0.35227273  0.11363636
  -0.05227273  0.125     ]
 [-0.22272727 -0.11818182 -0.21818182 -0.34772727 -0.01136364  0.43181818
  -0.18863636  0.125     ]
 [-0.00454545 -0.06363636  0.03636364 -0.12954545 -0.10227273 -0.11363636
   0.30227273  0.125     ]
 [-0.45       -0.3        -0.4        -0.575      -0.375      -0.25
  -0.325       0.125     ]]


The solution are rankings:


matrix([[12.00681818],
        [-0.92954545],
        [-4.72954545],
        [ 3.00681818],
        [-6.98409091],
        [-7.02954545],
        [ 5.73409091],
        [-0.075     ]])